internal/crd/crd_test.go

```diff
@@ -0,0 +1,181 @@
+			errString:         "cannot remove stored versions",
+			existingCrVersion: "v1alpha1",
+			existingCrdVersions: []apiextensionsv1.CustomResourceDefinitionVersion{
+				{
+					Name:    "v1alpha1",
+					Served:  true,
+					Storage: true,
+					Schema: &apiextensionsv1.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+							Type:        "object",
+							Description: "my crd schema",
+						},
+					},
+				},
+			},
+			newCrdVersions: []apiextensionsv1.CustomResourceDefinitionVersion{
+				{
+					Name:    "v1alpha2",
+					Served:  true,
+					Storage: true,
+					Schema: &apiextensionsv1.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+							Type:        "object",
+							Description: "my crd schema",
+						},
+					},
+				},
+			},
+		},
+		{
+			name:              "invalidates a crd upgrade that breaks an existing cr",
+			errString:         "failed validation for new schema",
+			existingCrVersion: "v1alpha1",
+			existingCrdVersions: []apiextensionsv1.CustomResourceDefinitionVersion{
+				{
+					Name:    "v1alpha1",
+					Served:  true,
+					Storage: true,
+					Schema: &apiextensionsv1.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+							Type:        "object",
+							Description: "my crd schema",
+							Properties: map[string]apiextensionsv1.JSONSchemaProps{
+								"sampleProperty": {Type: "string"},
+							},
+						},
+					},
+				},
+			},
+			newCrdVersions: []apiextensionsv1.CustomResourceDefinitionVersion{
+				{
+					Name:    "v1alpha1",
+					Served:  true,
+					Storage: true,
+					Schema: &apiextensionsv1.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+							Type:        "object",
+							Description: "my crd schema",
+							Required:    []string{"sampleProperty"},
+							Properties: map[string]apiextensionsv1.JSONSchemaProps{
+								"sampleProperty": {Type: "string"},
+							},
+						},
+					},
+				},
+			},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			defer ctx.Done()
+
+			// Create needed structs for testing using uniquely generated names
+			existingCrd := util.NewTestingCRD("samplecrd", "e2e.io", true, tt.existingCrdVersions)
+			uniqueName := existingCrd.Spec.Names.Singular
+			existingCr := util.NewTestingCR("samplecr", "e2e.io", tt.existingCrVersion, uniqueName)
+			newCrd := util.NewTestingCRD(uniqueName, "e2e.io", false, tt.newCrdVersions)
+
+			// Create existing CRD
+			err := kubeclient.Create(ctx, existingCrd)
+			require.NoError(t, err, "failed to create initial crd for testing")
+			require.Eventuallyf(t, func() bool {
+				return kubeclient.Get(ctx, client.ObjectKeyFromObject(existingCrd), existingCrd) == nil
+			}, defaultWaitPeriod, defaultTick, "failed to get initial crd for testing: %v", err)
+
+			// Creating existing CR
+			err = kubeclient.Create(ctx, existingCr)
```

Comment:
Since `envtest` actually uses an api-server under the hood to create resources, this test may be racy where the CRD is not accepted and established on the cluster (but is GET-able) by the time the CR is created. See https://github.com/operator-framework/operator-lifecycle-manager/blob/c36183ef17f757dbf1f96a2619a2a91184afc579/test/e2e/bundle_e2e_test.go#L179 for an example of a function that ensures the CRD is available on the cluster and ready to accept CRs. We might want something like this in our codebase too.

Reply:
Great point. I think we could benefit from a function like this in the `testutil` package that this PR adds too.